lib/agent.js

```diff
@@ -27,7 +32,78 @@ module.exports = class HTTPAgent {
   }
 
   keepSocketAlive (socket) {
-    return false
+    if (this._keepAlive === -1) return false
+
+    socket.setKeepAlive(true, this._keepAlive)
+    socket.unref()
+
+    return true
+  }
+
+  getName (opts) {
+    return `${opts.host}:${opts.port}`
+  }
+
+  addRequest (req, opts) {
+    opts = { ...opts, ...this._opts }
+
+    const name = this.getName(opts)
+
+    let socket
+
+    if (this._freeSockets.has(name)) {
+      socket = this._freeSockets.get(name).shift()
+      if (this._freeSockets.get(name).length === 0) this._freeSockets.delete(name)
+
+      this.reuseSocket(socket, req)
+    } else {
+      socket = this.createConnection(opts)
+
+      socket.on('free', () => this._onfree(socket, name))
+      socket.on('close', () => this._onremove(socket, name))
+    }
+
+    if (!this._sockets.has(name)) this._sockets.set(name, new Set())
+
+    this._sockets.get(name).add(socket)
+
+    req._onSocket(socket, opts)
+  }
+
+  destroy () {
+    const sets = [this._sockets, this._freeSockets]
+
+    sets.forEach((set) => {
+      set.forEach((sockets) => {
+        sockets.forEach((socket) => socket.destroy())
+      })
+    })
+  }
+
+  _onfree (socket, name) {
+    if (this.keepSocketAlive(socket)) {
+      this._onremove(socket, name, false) // remove from agent._sockets
+
+      if (!this._freeSockets.has(name)) this._freeSockets.set(name, [])
+      this._freeSockets.get(name).push(socket)
+    } else {
+      socket.end()
+    }
+  }
+
+  _onremove (socket, name, all = true) {
+    if (this._sockets.has(name)) {
+      this._sockets.get(name).delete(socket)
+
+      if (this._sockets.get(name).size === 0) this._sockets.delete(name)
+    }
+
+    if (all && this._freeSockets.has(name)) {
+      const index = this._freeSockets.get(name).indexOf(socket)
+      if (index !== -1) this._freeSockets.get(name).splice(index, 1)
```

Comment:
Ah, we of course also need to remove particular sockets from the free list. Let's make it a set to avoid the linear scan.